3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/OpenCageEnrichment.scala

```diff
@@ -0,0 +1,223 @@
+          geocode  <- getGeocodeFromResponse(response)
+          obj      <- geocodeToJValue(geocode)
+        } yield obj
+
+      case _ => s"One of the required event fields missing. latitude $latitude, longitude: $longitude".fail
+    }
+
+  private def geocodeToJValue(location: Geocode): Validation[String, JObject] =
+    Extraction.decompose(location) match {
+      case obj: JObject => obj.success
+      case _            => s"Couldn't transform location object $location into JSON".fail // Shouldn't ever happen
+    }
+
+  /**
+   * Return geocoding
+   */
+  private def getCachedOrRequest(latitude: Float, longitude: Float): Validation[String, OpenCageResponse] =
+    cache.getCachedOrRequest(latitude, longitude) match {
+      case Right(location) => location.success
+      case Left(error)     => error.getMessage.failure
+    }
+
+  /**
+   * Add Iglu URI to JSON Object
+   *
+   * @param context Opencage context as JSON Object
+   * @return JSON Object wrapped as Self-describing JSON
+   */
+  private def addSchema(context: JObject): JObject =
+    ("schema", schemaUri) ~ (("data", context))
+
+  /**
+   * From [[OpenCageResponse]] obtain the most relevant result
+   * for further JSON Object decomposition
+   *
+   * @param origin original OpenCageResponse object
+   * @return Most relevant Geocode location
+   */
+  private[enrichments] def getGeocodeFromResponse(origin: OpenCageResponse): Validation[String, Geocode] =
+    origin.results.headOption
+      .toSuccess("Empty list of results although the call went OK")
+
+}
+
+private[enrichments] case class TransformedLocation(address: String)
+
+case class OpenCageCache(
+  client: OpenCageClient,
+  cacheSize: Int, // Size of the LRU cache
+  scale: Int, // numbers of decimals to keep from latitudes and longitudes
+  timeout: Int // Number of seconds we wait for a response from Opencage
+) {
+
+  private val secondsTimeout = timeout.seconds
+
+  case class CacheKey(latitude: Float, longitude: Float)
+
+  type Cache = SynchronizedLruMap[CacheKey, Either[OpencageClientError, OpenCageResponse]]
+  protected val cache: Cache = new Cache(cacheSize)
+
+  def getCachedOrRequest(latitude: Float, longitude: Float): Either[OpencageClientError, OpenCageResponse] = {
+    val key = this.coordToCacheKey(latitude, longitude)
+    cache.get(key) match {
+      case Some(Right(cached))         => Right(cached)
+      case Some(Left(TimeoutError(_))) => getAndCache(latitude, longitude)
+      case Some(Left(error))           => Left(error)
+      case None                        => getAndCache(latitude, longitude)
+    }
+  }
+
+  private def getAndCache(latitude: Float, longitude: Float): Either[OpencageClientError, OpenCageResponse] = {
+    val apiCallResult = Try(Await.result(client.reverseGeocode(latitude, longitude), secondsTimeout)) match {
+      case util.Success(value) => Right(value)
+      case util.Failure(ex)    => Left(ex.asInstanceOf[OpencageClientError])
```

Comment:
why do we need the `asInstanceOf`/

Reply:
Failure of a future returns a Throwable. I want to cast it to the original type.

Reply:
Fixed.